### PR TITLE
WIP: gotosocial: init at 0.7.0

### PR DESCRIPTION
**Pull request by pbsds**
###### Description of changes

This PR packages [GotoSocial](https://gotosocial.org/), and adds a nixos module to set it up.

Current status: **draft**
TODOs:

* [ ] Fix the crash+stack trace that happens when running as `nixos-rebuild build-vm ...`
* [ ] Properly configure the db
* [ ] Add a nixos test
* [ ] Add some mechanism to create an admin user during first-time-setup

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/gotosocial.nix [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        };
+      };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = databaseActuallyCreateLocally -> (cfg.user == cfg.database.user);
+        message = ''For local automatic database provisioning (services.mastodon.database.createLocally == true) with peer authentication (services.mastodon.database.host == "/run/postgresql") to work services.mastodon.user and services.mastodon.database.user must be identical.'';
```
Comment (fleaz) [inline nixos/modules/services/web-apps/gotosocial.nix:140]: The usage of `mastodon` here looks like a copy/paste error and should probably be `gotosocial` I guess?

File: pkgs/servers/gotosocial/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{ lib
+, stdenv
+, buildGoModule
+, fetchFromGitHub
+}:
+
+buildGoModule rec {
+  pname = "gotosocial";
+  version = "0.5.2";
```
Comment (phaer) [inline pkgs/servers/gotosocial/default.nix:9]: Upstream has meanwhile published 0.7.0

**Review by phaer — CHANGES_REQUESTED** [review]
Besides @fleaz comments, this now needs an update but otherwise looks good to me

**Comment by pbsds** [comment]
I have yet to setup the nixos module properly w.r.t state. i find my current direction lacking. My time is spread thin, so should i remove the nixos module commit and just merge gotosocial as a package alone?

**Comment by abhin4v** [comment]
Maybe this will help wrt. setting up the module: https://notes.abhinavsarkar.net/2022/gotosocial-on-nixos

I'm the author of that post.

**Comment by fleaz** [comment]
I'm planning on running GoToSocial myself (that's how I found this old PR). If you currently don't have the time to finish the module, I would be happy to jump in and give it a try :)

**Comment by pbsds** [comment]
I just pushed my newest version, go ahead and work from that. Frankly i kind of lost interest, but would love to see someone see this PE through

**Comment by softinio** [comment]
> I'm planning on running GoToSocial myself (that's how I found this old PR). If you currently don't have the time to finish the module, I would be happy to jump in and give it a try :)

Ditto!

File: pkgs/servers/gotosocial/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    homepage = "https://docs.gotosocial.org/";
+    description = "Fast, fun, ActivityPub server, powered by Go";
+    longDescription = ''
+      ActivityPub social network server, written in Golang.
+      You can keep in touch with your friends, post, read, and
+      share images and articles. All without being tracked or
+      advertised to! A light-weight alternative to Mastodon
+      and Pleroma, with support for clients!
+    '';
+    maintainers = with maintainers; [ pbsds ];
```
Comment (softinio) [inline pkgs/servers/gotosocial/default.nix:30]: @pbsds as a suggestion, based on your comment [here](https://github.com/NixOS/nixpkgs/pull/202110#issuecomment-1442428541) you may want to take @fleaz offer [here](https://github.com/NixOS/nixpkgs/pull/202110#issuecomment-1442033253)

Comment (pbsds) [reply pkgs/servers/gotosocial/default.nix:30]: Although not spelled out explicitly, i did give the go ahead. I uploaded my newest version with this in mind

Comment (fleaz) [reply pkgs/servers/gotosocial/default.nix:30]: Sorry for the delay, i fell off a ladder and broke my arm 🙈  

Will start to work on this soon.

Comment (pbsds) [reply pkgs/servers/gotosocial/default.nix:30]: Best wishes!

**Comment by misuzu** [comment]
Please take a look on my take on this: https://github.com/NixOS/nixpkgs/pull/232862

**Comment by pbsds** [comment]
Far better than mine.

**Pull request closed without merging**